concourse/steps/release.py

```diff
@@ -576,21 +576,25 @@ def name(self):
 
     def validate(self):
         version.parse_to_semver(self.release_version)
+
         # either cds _XOR_ ctf must exist
         have_ctf = os.path.exists(self.ctf_path)
         have_cd = os.path.exists(self.component_descriptor_v2_path)
         if not have_ctf ^ have_cd:
             ci.util.fail('exactly one of component-descriptor, or ctf-archive must exist')
         elif have_cd:
-            component_descriptor_v2 = cm.ComponentDescriptor.from_dict(
+            self.components = [cm.ComponentDescriptor.from_dict(
                 ci.util.parse_yaml_file(self.component_descriptor_v2_path),
-            )
-            # resolve all referenced components (thus ensure all dependencies are available)
-            tuple(cnudie.retrieve.components(component=component_descriptor_v2))
-            # TODO: more validation (e.g. check for uniqueness of names)
+            )]
         elif have_ctf:
-            # nothing to do, already uploaded in component_descriptor step.
-            pass
+            self.components = list(component_descriptors_from_ctf_archive(
+                self.ctf_path,
+            ))
+            if not component_descriptors:
+                ci.util.fail(f'No component descriptor found in CTF archive at {self.ctf_path}')
+
+        for component_descriptor_v2 in self.components:
+            tuple(cnudie.retrieve.components(component=component_descriptor_v2))
```

Comment:
nit: since you merely want to exhaust the generator (at least this is what I assume), a nested `for` might be a bit more explicit

Reply:
not sure about the nested for.
From my understanding the code just checks if all components can be resolved. so we do not even need a generator?

Reply:
the generator is returned from the `cnudie.retrieve.components` function (implicitly, because its definition contains `yield` stmts).

generators are evaluated lazily. that means, that if they are not consumed, no code will be executed at all. creating a `tuple` is a valid means to for the generator to be consumed (so your code is perfectly correct).

however, this has two disadvantages:

1. creation of the tuple will require memory (and thus destroys the potential benefit from the generator)
2. it may not be obvious to other readers that the tuple is created for the sole purpose of exhausting the generator

thus I think it might be a better "pattern" to use a nested for-loop, like so: